zkml/src/layers/provable/mod.rs

```diff
@@ -0,0 +1,502 @@
+        // Default implementation, to avoid having to implement this method in case `is_provable` is false
+        assert!(
+            !self.is_provable(),
+            "Running default prove implementation for a provable operation! Implement prove method"
+        );
+        Ok(vec![Claim::default()])
+    }
+
+    /// Generate witness for a node where a lookup table is employed in proving
+    fn gen_lookup_witness(
+        &self,
+        _id: NodeId,
+        _gen: &mut LookupWitnessGen<E>,
+        _step_data: &StepData<Element, E>,
+    ) -> Result<(), ProvableOpError> {
+        // Default implementation for nodes that don't employ a lookup table
+        Ok(())
+    }
+}
+
+pub trait Op<E, N>: Evaluate<N> + ProveInfo<E> + Debug + ProvableOp<E>
+where
+    E: ExtensionField,
+    E::BaseField: Serialize + DeserializeOwned,
+    E: Serialize + DeserializeOwned,
+    N: Number,
+{
+}
+
+pub trait VerifiableCtx<E>: Debug
+where
+    E: ExtensionField,
+    E::BaseField: Serialize + DeserializeOwned,
+    E: Serialize + DeserializeOwned,
+{
+    type Proof: Sized;
+
+    /// Verify proof for the given operation
+    fn verify<T: Transcript<E>>(
+        &self,
+        proof: &Self::Proof,
+        last_claims: &[&Claim<E>],
+        verifier: &mut Verifier<E, T>,
+        shape_step: &ShapeStep,
+    ) -> Result<Vec<Claim<E>>, ProvableOpError>;
+
+    /// Compute the shapes of the output tensors, given the shapes of the input tensors
+    /// and the padding mode
+    fn output_shapes(
+        &self,
+        input_shapes: &[Vec<usize>],
+        padding_mode: PaddingMode,
+    ) -> Vec<Vec<usize>>;
+}
+
+// Helper method to call `VerifiableCtx::output_shapes` when the type `E`
+// cannot be inferred automatically by the compiler
+pub(crate) fn output_shapes<E: ExtensionField, C: VerifiableCtx<E>>(
+    ctx: &C,
+    input_shapes: &[Vec<usize>],
+    padding_mode: PaddingMode,
+) -> Vec<Vec<usize>>
+where
+    E::BaseField: Serialize + DeserializeOwned,
+    E: Serialize + DeserializeOwned,
+{
+    ctx.output_shapes(input_shapes, padding_mode)
+}
+
+impl<E: ExtensionField> VerifiableCtx<E> for LayerCtx<E>
+where
+    E::BaseField: Serialize + DeserializeOwned,
+    E: Serialize + DeserializeOwned,
+{
+    type Proof = LayerProof<E>;
+
+    fn verify<T: Transcript<E>>(
+        &self,
+        proof: &LayerProof<E>,
+        last_claims: &[&Claim<E>],
+        verifier: &mut Verifier<E, T>,
+        shape_step: &ShapeStep,
+    ) -> Result<Vec<Claim<E>>, ProvableOpError> {
+        match self {
+            LayerCtx::Dense(dense_ctx) => {
+                if let LayerProof::Dense(proof) = proof {
+                    dense_ctx.verify(proof, last_claims, verifier, shape_step)
+                } else {
+                    Err(ProvableOpError::ParameterError(
+                        "dense proof not found for dense layer".to_string(),
+                    ))
+                }
+            }
```

Comment:
Nothing to act on but this is exactly the kind of repetitive structure we could easily ask a proc macro to generate for us i believe.